server/init.go

```diff
@@ -19,203 +25,331 @@ import (
+			if err != nil {
+				return err
+			}
+
+			genFile := config.GenesisFile()
+			if !viper.GetBool(flagOverwrite) && cmn.FileExists(genFile) {
+				return fmt.Errorf("genesis config file already exists: %v", genFile)
+			}
+
+			err = WriteGenesisFile(cdc, genFile, chainID, validators, appState)
+			if err != nil {
+				return err
+			}
+
+			nodeKey, err := p2p.LoadOrGenNodeKey(config.NodeKeyFile())
+			if err != nil {
+				return err
+			}
+			nodeID := string(nodeKey.ID())
+
+			// print out some key information
+
+			toPrint := struct {
+				ChainID    string          `json:"chain_id"`
+				NodeID     string          `json:"node_id"`
+				AppMessage json.RawMessage `json:"app_message"`
+			}{
+				chainID,
+				nodeID,
+				cliPrint,
+			}
+			out, err := wire.MarshalJSONIndent(cdc, toPrint)
+			if err != nil {
+				return err
+			}
+			fmt.Println(string(out))
+
+			// write the piece file is path specified
+			if viper.GetBool(flagPieceFile) {
+				//create the piece
+				ip, err := externalIP()
+				if err != nil {
+					return err
+				}
+				piece := GenesisPiece{
+					ChainID:    chainID,
+					NodeID:     nodeID,
+					IP:         ip,
+					AppState:   appState,
+					Validators: validators,
+				}
+				bz, err := cdc.MarshalJSON(piece)
+				if err != nil {
+					return err
+				}
+				name := fmt.Sprintf("piece%v.json", nodeID)
+				file := filepath.Join(viper.GetString("home"), name)
+				return cmn.WriteFile(file, bz, 0644)
+			}
+
+			return nil
+		},
+	}
+	if appInit.AppendAppState != nil {
+		cmd.AddCommand(FromPiecesCmd(ctx, cdc, appInit))
+		cmd.Flags().BoolP(flagPieceFile, "a", false, "create an append file (under [--home]/[nodeID]piece.json) for others to import")
+	}
+	cmd.Flags().BoolP(flagOverwrite, "o", false, "overwrite the config file")
+	cmd.Flags().AddFlagSet(appInit.Flags)
+	return cmd
 }
 
-type initCmd struct {
-	genAppState GenAppState
-	context     *Context
+// genesis piece structure for creating combined genesis
+type GenesisPiece struct {
+	ChainID    string                     `json:"chain_id"`
+	NodeID     string                     `json:"node_id"`
+	IP         string                     `json:"ip"`
+	AppState   json.RawMessage            `json:"app_state"`
+	Validators []tmtypes.GenesisValidator `json:"validators"`
 }
 
-// InitCmd will initialize all files for tendermint,
-// along with proper app_state.
-// The application can pass in a function to generate
-// proper state. And may want to use GenerateCoinKey
-// to create default account(s).
-func InitCmd(gen GenAppState, ctx *Context) *cobra.Command {
-	cmd := initCmd{
-		genAppState: gen,
-		context:     ctx,
-	}
-	cobraCmd := cobra.Command{
-		Use:   "init",
-		Short: "Initialize genesis files",
-		RunE:  cmd.run,
+// get cmd to initialize all files for tendermint and application
+func FromPiecesCmd(ctx *Context, cdc *wire.Codec, appInit AppInit) *cobra.Command {
```

Comment:
Is the idea that each node will generate a genesis file, then send them to each other and run this command to combine them?

That could have security benefits over coordinating with a web tool, but I'm not sure it would be easier for testing setups, seems like it would require a lot of communication (now `n - 1` files have to be sent between `n` peers, instead of one leader assembling `genesis.json` in a text editor).

I guess it's always better to have the option - but the webapp approach ([alpha](https://github.com/tendermint/alpha)) seems like it might be more flexible / usable.

Reply:
> Is the idea that each node will generate a genesis file, then send them to each other and run this command to combine them?

something like that - we don't distribute your genesis though, but a new "piece" json file

> That could have security benefits over coordinating with a web tool, 

it does

> but I'm not sure it would be easier for testing setups, seems like it would require a lot of communication (now n - 1 files have to be sent between n peers, instead of one leader assembling genesis.json in a text editor).

I disagree - it's much easier - as from previous discussions, you just make a PR to a repo to include your genesis piece - then everyone pulls from the repo - v easy - if you want to make it even easier we could have a web tool where anyone can upload a file to a folder and then the tool serves out the folder - I think github is good enough personally

> The webapp approach (alpha) seems like it might be more flexible / usable for simple testnets. This seems like a good solution for actual setups of distributed validator sets, if we make sure it's reliable and outline the workflow clearly.

I disagree, now that there is a simple solution for coordination between testnet participants,  I don't think we want/need testnet-setup-webtools (faucet web-tools are still great) anymore / we shouldn't waste time maintain them as we rapidly iterate to the final design. By keeping development in the SDK we create a maintainable situation where as the needs of genesis change the developer upgrading the genesis also upgrades the testnet tooling. The only tool we need is the CLI and github

btw this idea is based on feedback from the validator group - couldn't agree more with this approach. lol sorry if I sound defensive, but I do have a strong perspective on this.

Reply:
I agree completely with maintainability and direct inclusion in the SDK.

If this is based directly on validator group feedback, by all means we should do it - I thought previous attempts to coordinate genesis files with Github were unsuccessful, but maybe that was because of merge conflicts to a single file, which this would solve.

Reply:
Yeah previously there was just one file. Now you just open PR with the addition of your new "piece" file so there are no merge conflicts here - it can actually be done quite easily directly within the github website so it seems easy enough - I think the only thing that would be better would be to deploy a server which the `gaiad init` could point to to directly send the output, and which would be able to serve up all the pieces

Reply:
Possible to automate through Git with a repository URL - https://github.com/src-d/go-git- maybe overkill.